### PR TITLE
Refactor RootCoord

### PR DESCRIPTION
**Pull request by longjiquan**
issue: #17412 
Signed-off-by: longjiquan <jiquan.long@zilliz.com>
Co-authored-by: xaxys <tpnnghd@163.com>

### CODE REVIEW

File: internal/proto/etcd_meta.proto [hunk 1]
```diff
@@ -20,6 +20,27 @@ message FieldIndexInfo{
   int64 indexID = 2;
 }
 
+enum CollectionState {
+  CollectionCreated = 0;
```
Comment (czs007) [inline internal/proto/etcd_meta.proto:24]: Add CollectionNone=0

**Comment by longjiquan** [comment]
rerun ut

**Comment by longjiquan** [comment]
rerun ut

File: internal/common/map.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package common
+
+import "reflect"
+
+type S2S map[string]string
```
Comment (czs007) [inline internal/common/map.go:5]: S2S ==> Str2Str

File: internal/metastore/db/rootcoord/table_catalog.go [hunk 3]
```diff
@@ -375,6 +379,63 @@ func (tc *Catalog) DropCollection(ctx context.Context, collection *model.Collect
 	})
 }
 
+func (tc *Catalog) alterModifyCollection(ctx context.Context, oldColl *model.Collection, newColl *model.Collection, ts typeutil.Timestamp) error {
+	if oldColl.TenantID != newColl.TenantID || oldColl.CollectionID != newColl.CollectionID {
+		return fmt.Errorf("altering tenant id or collection id is forbidden")
+	}
+	oldCollClone := oldColl.Clone()
+	newCollClone := newColl.Clone()
```
Comment (czs007) [inline internal/metastore/db/rootcoord/table_catalog.go:387]: unnecessary clone

File: internal/metastore/kv/rootcoord/kv_catalog.go [hunk 4]
```diff
@@ -334,39 +366,81 @@ func (kc *Catalog) AlterAlias(ctx context.Context, alias *model.Alias, ts typeut
...
+	// if we found collection dropping, we should try removing related resources.
+	return kc.Snapshot.MultiSaveAndRemoveWithPrefix(nil, []string{collectionKey}, ts)
+}
+
+func (kc *Catalog) alterModifyCollection(oldColl *model.Collection, newColl *model.Collection, ts typeutil.Timestamp) error {
+	if oldColl.TenantID != newColl.TenantID || oldColl.CollectionID != newColl.CollectionID {
+		return fmt.Errorf("altering tenant id or collection id is forbidden")
+	}
+	oldCollClone := oldColl.Clone()
+	newCollClone := newColl.Clone()
```
Comment (czs007) [inline internal/metastore/kv/rootcoord/kv_catalog.go:396]: uncessary clone?

File: internal/metastore/model/field.go [hunk 5]
```diff
@@ -14,6 +18,69 @@ type Field struct {
...
+
+func (f Field) Clone() *Field {
+	return &Field{
+		FieldID:      f.FieldID,
+		Name:         f.Name,
+		IsPrimaryKey: f.IsPrimaryKey,
+		Description:  f.Description,
+		DataType:     f.DataType,
+		TypeParams:   common.CloneKeyValuePairs(f.TypeParams),
+		IndexParams:  common.CloneKeyValuePairs(f.IndexParams),
```
Comment (czs007) [inline internal/metastore/model/field.go:36]: actually , this IndexParams is banned. no need to save ?

File: internal/rootcoord/undo.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		undoStep: make([]Step, 0),
+	}
+}
+
+func (b *baseUndoTask) AddStep(todoStep, undoStep Step) {
+	b.todoStep = append(b.todoStep, todoStep)
+	b.undoStep = append(b.undoStep, undoStep)
+}
+
+func (b *baseUndoTask) undoUntilLastFinished(lastFinished int) {
```
Comment (czs007) [inline internal/rootcoord/undo.go:29]: better to renamed as undoFromLastFinished

File: internal/rootcoord/meta_table.go [hunk 7]
```diff
@@ -64,809 +68,590 @@ const (
...
-	collMeta, ok := mt.collID2Meta[collID]
-	if !ok {
-		return model.Collection{}, fmt.Errorf("collection not found: %s", collName)
-	}
-	return collMeta, nil
+	return mt.listAliasesByID(collID)
 }
 
-//func (mt *MetaTable) checkFieldCanBeIndexed(collMeta model.Collection, fieldSchema model.Field, idxInfo *model.Index) error {
-//	for _, tuple := range collMeta.FieldIDToIndexID {
```
Comment (czs007) [inline internal/rootcoord/meta_table.go:572]: just remove this.

File: internal/rootcoord/root_coord.go [hunk 8]
```diff
@@ -1207,481 +689,505 @@ func (c *Core) GetStatisticsChannel(ctx context.Context) (*milvuspb.StringRespon
...
-//
-//	metrics.RootCoordDDLReqCounter.WithLabelValues("DescribeSegments", metrics.SuccessLabel).Inc()
-//	metrics.RootCoordDDLReqLatency.WithLabelValues("DescribeSegments").Observe(float64(tr.ElapseSpan().Milliseconds()))
-//
-//	t.Rsp.Status = succStatus()
-//	return t.Rsp, nil
-//}
-
 // ShowSegments list all segments
 func (c *Core) ShowSegments(ctx context.Context, in *milvuspb.ShowSegmentsRequest) (*milvuspb.ShowSegmentsResponse, error) {
```
Comment (czs007) [inline internal/rootcoord/root_coord.go:1187]: Is this still in use?

**Comment by longjiquan** [comment]
rerun ut

**Pull request merged**